storage_mon: Use the O_DIRECT flag in open() to eliminate cache effects

inouekazu opened the pull request:
Use of O_DIRECT in open() eliminates the possibility of false negatives resulting from buffer cache reads.

please refer to https://lists.clusterlabs.org/pipermail/users/2022-August/030459.html for background.

fabbione commented:
test this please

chrissie-c commented:
I agree with Ken (in the above thread) that there should be a fallback if O_DIRECT fails. There are filesystems and options even on Linux that don't always support O_DIRECT and they shouldn't be ecxluded or errored

inouekazu commented:
- https://github.com/ClusterLabs/resource-agents/pull/1797/commits/ce4e632f29ed6b86b82a959eac5844655baed153
   - According to https://ci.kronosnet.org/job/resource-agents-build-all-voting/1530/, build targets are FreeBSD and RHEL etc., so I deleted ```#if defined(__linux__) || defined(__FreeBSD__)```.
   - Fixed FreeBSD build failure.
 - The open() manual says it returns EINVAL for filesystems that do not support O_DIRECT. Are there any issues that the implementation below does not address?
   - https://github.com/ClusterLabs/resource-agents/pull/1797/files#diff-91e236cc5ccce26f5a934e98aef5f1082251a8676e999b5364add5603538483dR44

chrissie-c reviewed (changes requested):
There are few things I'm not clear/happy about still.

Is there a reason that the random lseek doesn't happen with O_DIRECT? I must admit I haven't looked into it in detail but it seems odd that the (hopefully) majority test case only checks the start of the device.

Also, I don't understand why the error checking for the ioctl is done on the sec_size value rather than the ioctl return. If that's fine then a comment saying why would be helpful. Maybe we need both?

In the case where O_DIRECT is not being used, we can skip the alignment calls (ioctl & posix_memalign) - they are just another potential cause of failure which we can do without - we're looking for disk errors here so the fewer syscalls that can fail the better.

wenningerk commented:
> 
> Also, I don't understand why the error checking for the ioctl is done on the sec_size value rather than the ioctl return. If that's fine then a comment saying why would be helpful. Maybe we need both?
> 
Some kind of shortcut for 'ioctl has to succeed and not return bogus'.
But some comment would of course make sense.
Same code-pattern btw. has been in sbd since ever.

inouekazu commented:
- https://github.com/ClusterLabs/resource-agents/pull/1797/commits/7a0aaa0dfdebeab3fae9fe9ddc412c3d1f610273
   - reflected the https://github.com/ClusterLabs/resource-agents/pull/1797#pullrequestreview-1080084034.

chrissie-c commented:
retest this please

chrissie-c commented:
retest this please

wenningerk commented on tools/storage_mon.c:
```diff
@@ -72,65 +77,58 @@ static int open_device(const char *device, int verbose)
-		goto error;
-	}
+		if (res < 0) {
+			fprintf(stderr, "Failed to stat %s: %s\n", device, strerror(errno));
```
why don't we simply say that getting the block-size did fail?

inouekazu replied:
I agree. https://github.com/ClusterLabs/resource-agents/pull/1797/commits/db97e055a17526cec056c595844a9d8851e3ee19

wenningerk commented on tools/storage_mon.c:
```diff
@@ -72,65 +77,58 @@ static int open_device(const char *device, int verbose)
-		goto error;
-	}
+		if (posix_memalign(&buffer, sysconf(_SC_PAGESIZE), sec_size) != 0) {
+			fprintf(stderr, "Failed to allocate aligned memory: %s\n", strerror(errno));
```
What I've read alignment should be more block-size. That makes sense thinking of a contiguous
block of memory cut into block-size chunks that you would like to be able to pass in any combination.
Usually it will probably be 512 bytes block-size vs. 4k pagesize and thus Ok.
Block-size of many disks being 4k nowadays I would expect that to become larger in future.
(or already in special cases like maybe raids, ssds with their large erase-block-size, shingled-write disks, ...)
max(block-size, pagesize) might be a way to satisfy everybody as much as possible.

wenningerk replied:
sbd is using block-size - at least since I've reworked that code-part a bit but iirc that was there since ever.

wenningerk replied:
aah - crap you aren't using aio here. So maybe forget what I've said.

oalbrigt commented:
Thanks.
